#! /bin/sh /usr/share/dpatch/dpatch-run
## app_queue.debian.leakfix.patch by Kiss Karoly <karcsi@tvnetwork.hu>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: backports memleak fix http://bugs.digium.com/view.php?id=4318

@DPATCH@
--- apps/app_queue.c.orig	2005-08-30 10:55:15.000000000 +0200
+++ apps/app_queue.c	2005-08-30 10:55:24.000000000 +0200
@@ -861,21 +861,30 @@
 			if (!f || ((f->frametype == AST_FRAME_CONTROL) && (f->subclass == AST_CONTROL_HANGUP))) {
 				/* Got hung up */
 				*to=-1;
+				if ( f )
+					ast_frfree(f);
 				return NULL;
 			}
-			if (f && (f->frametype == AST_FRAME_DTMF) && allowdisconnect_out && (f->subclass == '*')) {
+			if ((f->frametype == AST_FRAME_DTMF) && allowdisconnect_out && (f->subclass == '*')) {
 			    if (option_verbose > 3)
 					ast_verbose(VERBOSE_PREFIX_3 "User hit %c to disconnect call.\n", f->subclass);
 				*to=0;
+				if ( f )
+					ast_frfree(f);
 				return NULL;
 			}
-			if (f && (f->frametype == AST_FRAME_DTMF) && (f->subclass != '*') && valid_exit(qe, f->subclass)) {
+			if ((f->frametype == AST_FRAME_DTMF) && (f->subclass != '*') && valid_exit(qe, f->subclass)) {
 				if (option_verbose > 3)
 					ast_verbose(VERBOSE_PREFIX_3 "User pressed digit: %c", f->subclass);
 				*to=0;
 				*digit=f->subclass;
+				if ( f )
+					ast_frfree(f);
 				return NULL;
 			}
+			if ( f )
+				ast_frfree(f);
+
 		}
 		if (!*to && (option_verbose > 2))
 			ast_verbose( VERBOSE_PREFIX_3 "Nobody picked up in %d ms\n", orig);
